### PR TITLE
[SQUALL-25] Change compression approach

### LINKED ISSUE #25: Change compression approach
We should remove GZip middleware and will make built-in, switchable compression based on ISA-L. 

https://github.com/intel/isa-l

### PR DESCRIPTION
**Pull request by naviot**
Closes #25

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -424,6 +425,25 @@ async def handle_get():
     ]
 ```
 
+## Compression
+
```
Comment (stas-dubrovskyi) [inline README.md:429]: @naviot More passion ))))

Squall provides built-in blazing-fast compression based on Intel® Intelligent Storage Acceleration Library (Intel® ISA-L) using awesome Python's [isal](https://pypi.org/project/isal/) library as binding. 

Such in-app performance does game-changing opportunities for the entire system set up, read more here ....

File: README.md [hunk 2]
```diff
@@ -37,6 +37,7 @@
   - [Body processing](#body-processing)
     - [Response serialization](#response-serialization)
     - [Response deserialization serialization](#response-deserialization-serialization)
+- [Compression](#compression)
```
Comment (stas-dubrovskyi) [inline README.md:40]: Seems like the wrong level. Should be nested in Usage

File: squall/responses.py [hunk 3]
```diff
@@ -74,16 +77,35 @@ def __init__(
...
+            and len(send_body["body"]) > compression.minimum_size
+        ):
+            accept_encoding = self.request.headers.get("Accept-Encoding", "")
+            for backend in compression.backends:
+                if backend.encoding_name in accept_encoding:
+                    body = backend.compress(send_body["body"], compression.level)
+                    headers = MutableHeaders(raw=self.send_start["headers"])
+                    headers["Content-Encoding"] = backend.encoding_name
+                    headers["Content-Length"] = str(len(body))
+                    headers.add_vary_header("Accept-Encoding")
```
Comment (stas-dubrovskyi) [inline squall/responses.py:103]: Something strange here. 
Headers assign is not explicit

File: tests/test_compression.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+def test_compression_response_classes(response_class, response, expected_response):
+    app = Squall(compression=Compression(minimum_size=1))
+
+    @app.get("/", response_class=response_class)
+    def get_root():
+        return response
+
+    client = TestClient(app)
+    response = client.get("/")
+    assert response.content == expected_response
```
Comment (stas-dubrovskyi) [inline tests/test_compression.py:45]: Missed assert for Content-Encoding

File: squall/responses.py [hunk 5]
```diff
@@ -74,16 +77,35 @@ def __init__(
...
+        self.send_start: Dict[str, Any] = {
             "type": "http.response.start",
             "status": status_code,
             "headers": raw_headers,
         }
         self.send_body = {"type": "http.response.body", "body": body}
 
     async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
+        compression: Optional[Compression] = scope["app"].compression
+        send_body: Dict[str, Any] = self.send_body
```
Comment (stas-dubrovskyi) [inline squall/responses.py:89]:
```
body = self.send_body.get('body', '')
```
Seems like there are we are saving some attributes accessing, but do more dictionary lookups. So it's like  vice-versa, but less readable

File: tests/test_compression.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+@pytest.mark.parametrize(
+    "backends",
+    [
+        [GzipBackend(), ZlibBackend()],
+        [ZlibBackend()],
+        [GzipBackend()],
+    ],
+)
+def test_compression_backends(backends):
+    app = Squall(compression=Compression(backends=backends, minimum_size=1))
```
Comment (stas-dubrovskyi) [inline tests/test_compression.py:17]: minimal

Comment (naviot) [reply tests/test_compression.py]: fixed

File: squall/responses.py (hunk 5 above)
Comment (naviot) [reply squall/responses.py]: changed

File: tests/test_compression.py (hunk 4 above)
Comment (naviot) [reply tests/test_compression.py:46]: done

File: squall/responses.py (hunk 3 above)
Comment (naviot) [reply squall/responses.py:94]: can't change right now
because initially self.send_start["headers"] is list of tuples

File: README.md (hunk 2 above)
Comment (naviot) [reply README.md]: fixed

File: README.md (hunk 1 above)
Comment (naviot) [reply README.md]: update doc

**Review by stas-dubrovskyi — APPROVED** [review]
🥇

**Pull request merged**